### ISSUE TITLE
Multiple calls to NS.compiled_definitions.get_parsed_defs()

### ISSUE THREAD

**Issue by r0h4n** [Tendrl/commons#586]
Multiple calls to NS.compiled_definitions.get_parsed_defs() in Tendrl code parses compiled definitions (yaml) every time said function is called

These calls cumulatively take the most amount of time from data gather using 
- https://github.com/ajdavis/GreenletProfiler
- Function stats in "pstats" format for python

```
Sun Jun 11 00:47:27 2017    profile.pstat

         211546888 function calls (213507390 primitive calls) in 1214.751 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    27/28  486.982   18.036 7426.608  265.236 /usr/lib64/python2.7/site-packages/gevent/_threading.py:414(Queue.get)
    27/28    0.001    0.000 6757.277  241.331 /usr/lib64/python2.7/site-packages/gevent/_threading.py:139(Condition.wait)
       28    0.000    0.000 1402.000   50.071 /usr/lib64/python2.7/site-packages/gevent/threadpool.py:325(wrap_errors)
      997    0.065    0.000  545.577    0.547 /usr/lib/python2.7/site-packages/tendrl/node_agent/objects/compiled_definition/__init__.py:21(CompiledDefinitions.get_parsed_defs)
      997    0.029    0.000  545.512    0.547 /usr/lib64/python2.7/site-packages/ruamel/yaml/main.py:99(safe_load)
      997    1.224    0.001  545.483    0.547 /usr/lib64/python2.7/site-packages/ruamel/yaml/main.py:73(load)
      997    0.007    0.000  543.669    0.545 /usr/lib64/python2.7/site-packages/ruamel/yaml/constructor.py:52(SafeLoader.get_single_data)
      970    0.011    0.000  534.405    0.551 /usr/lib/python2.7/site-packages/tendrl/commons/objects/__init__.py:19(TendrlContext.__init__)
      970    0.027    0.000  534.377    0.551 /usr/lib/python2.7/site-packages/tendrl/commons/objects/__init__.py:27(TendrlContext.load_definition)
      970    0.043    0.000  531.106    0.548 /usr/lib/python2.7/site-packages/tendrl/commons/__init__.py:374(TendrlNS.get_obj_definition)
      997    0.012    0.000  513.608    0.515 /usr/lib64/python2.7/site-packages/ruamel/yaml/composer.py:43(SafeLoader.get_single_node)
      997    0.008    0.000  513.054    0.515 /usr/lib64/python2.7/site-packages/ruamel/yaml/composer.py:65(SafeLoader.compose_document)
997/1457614    9.794    0.010  513.026    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/composer.py:78(SafeLoader.compose_node)
997/234295    5.600    0.006  512.812    0.002 /usr/lib64/python2.7/site-packages/ruamel/yaml/composer.py:146(SafeLoader.compose_mapping_node)
  4412722   10.900    0.000  471.859    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:112(SafeLoader.check_event)
   705876    5.270    0.000  254.201    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:514(SafeLoader.parse_block_mapping_value)
  1954120   14.596    0.000  239.027    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:163(SafeLoader.fetch_more_tokens)
      363    0.016    0.000  204.302    0.563 /usr/lib/python2.7/site-packages/tendrl/commons/objects/__init__.py:303(TendrlContext._copy_vars)
  3661981    8.218    0.000  203.587    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:135(SafeLoader.peek_token)
      336    0.014    0.000  189.424    0.564 /usr/lib/python2.7/site-packages/tendrl/commons/objects/node_context/__init__.py:23(NodeContext.__init__)
   940171    7.710    0.000  177.028    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:490(SafeLoader.parse_block_mapping_key)
 11565200   33.994    0.000  167.989    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:123(SafeLoader.check_token)
  1198394    4.624    0.000  150.999    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:681(SafeLoader.fetch_plain)
  1198394   34.269    0.000  135.858    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:1323(SafeLoader.scan_plain)
      243    0.005    0.000  135.213    0.556 /usr/lib/python2.7/site-packages/tendrl/commons/objects/service/__init__.py:8(Service.__init__)
  1457614   18.818    0.000  118.858    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:301(SafeLoader.parse_node)
  1411752    2.666    0.000  117.904    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:295(SafeLoader.parse_block_node_or_indentless_sequence)
 20310884   49.855    0.000   92.824    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:152(SafeLoader.need_more_tokens)
  1574263   21.411    0.000   56.103    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:1379(SafeLoader.scan_plain_spaces)
      104    0.003    0.000   55.019    0.529 /usr/lib/python2.7/site-packages/tendrl/commons/objects/block_device/__init__.py:5(BlockDevice.__init__)
       78    0.001    0.000   42.130    0.540 /usr/lib/python2.7/site-packages/tendrl/commons/objects/node_network/__init__.py:7(NodeNetwork.__init__)
 11439578   33.086    0.000   38.862    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/reader.py:108(SafeLoader.forward)
  1954120   14.404    0.000   36.641    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:764(SafeLoader.scan_to_next_token)
 42109292   35.164    0.000   35.164    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/reader.py:96(SafeLoader.peek)
   234295    0.853    0.000   32.077    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/parser.py:485(SafeLoader.parse_block_mapping_first_key)
 21007787   30.343    0.000   30.343    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:289(SafeLoader.stale_possible_simple_keys)
      997    0.268    0.000   30.054    0.030 /usr/lib64/python2.7/site-packages/ruamel/yaml/constructor.py:59(SafeLoader.construct_document)
235292/468590    0.796    0.000   29.010    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/constructor.py:491(SafeLoader.construct_yaml_map)
   234295    0.862    0.000   28.224    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/constructor.py:260(SafeLoader.construct_mapping)
       52    0.001    0.000   27.974    0.538 /usr/lib/python2.7/site-packages/tendrl/node_agent/objects/global_network/__init__.py:6(GlobalNetwork.__init__)
  3129583    8.526    0.000   26.030    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/scanner.py:142(SafeLoader.get_token)
   234295    2.837    0.000   25.607    0.000 /usr/lib64/python2.7/site-packages/ruamel/yaml/constructor.py:146(SafeLoader.construct_mapping)

```

PS:  currently we are optimizing "/site-packages/tendrl/" calls only, third party libs will be dealt with later

**Comment by r0h4n** [comment]
https://github.com/Tendrl/node-agent/commit/afa60cfa7e53a4364c457b8dbe18eea98755ea70

**Comment by r0h4n** [comment]
It seems this issue was increasing a lot of CPU time consumed by Tendrl, I ran below script by @dahorak for checking cpu/mem usage.
```
[root@k4n0-work ~]# cat tendrl_perf.sh 
date>>perf_tendrl_stat
while true
do

ps -p $(echo $(ps aux | grep [t]endrl-node-agent | awk '{print $2}') | sed 's/ /,/') -o %cpu,%mem,cmd -h>>perf_tendrl_stat

sleep 3;

done

```

**Develop branch:**
```
Sat Jun 10 23:16:40 IST 2017
64.6  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
40.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
42.0  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
33.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
29.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
25.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
22.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
23.6  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
23.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
23.0  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
22.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
23.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
23.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
22.1  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
20.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
21.3  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
21.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
20.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
19.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
18.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
18.0  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
17.1  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
16.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
16.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
15.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
15.3  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
14.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
14.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
13.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
13.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent


```

**With fix for Tendrl/commons#586**
```
Sun Jun 11 01:08:05 IST 2017
10.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
10.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
10.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 9.5  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 9.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 8.8  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 8.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 7.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 7.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 6.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 6.6  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 6.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 5.9  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 5.7  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 5.4  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 5.2  0.4 /usr/bin/python /usr/bin/tendrl-node-agent
 5.0  0.4 /usr/bin/python /usr/bin/tendrl-node-agent

```


These tests were run on below hardware:
`[root@k4n0-work ~]# lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    1
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 58
Model name:            Intel(R) Core(TM) i5-3470 CPU @ 3.20GHz
Stepping:              9
CPU MHz:               1603.250
BogoMIPS:              6385.47
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              6144K
NUMA node0 CPU(s):     0-3
`
